### ISSUE TITLE
Cannot Marshal.dump

### ISSUE THREAD

**Issue by maxime-carbonneau** [shioyama/mobility#494]
Will it be possible to Marshal.dump a AR that extend Mobility

## Context

config/initializers/mobility.rb
```ruby
Mobility.configure do 
  plugins do
    backend :table
    active_record
    reader
    writer
    backend_reader
    query
  end  
end
```

app/models/project.rb
```ruby
class Project < ActiveRecord::Base
  extend Mobility
  translates :title
end
```

## Actual Behavior

```ruby
p = Project.new(title: "My title")
Marshal.dump(p)
...
TypeError (can't dump hash with default proc)
```

**Comment by shioyama** [comment]
Hmm never tried that... I'll have to look into it. I assume here that `Project` inherits from `ActiveRecord::Base` right?

**Comment by maxime-carbonneau** [comment]
> Hmm never tried that... I'll have to look into it. I assume here that `Project` inherits from `ActiveRecord::Base` right?

Yes. The initial post is updated

**Comment by shioyama** [comment]
Sorry I can't reproduce this, `Marshal.dump(p)` works for me (no exception).

Are you sure there isn't something else involved?

**Comment by maxime-carbonneau** [comment]
Just to be sure, I start a new Rails project with :
```Ruby 2.7.2```
```Rails 6.1.1```
```rails new my_project --database postgresql```

I only add your gem to my ```Gemfile```
I use ```config/initializers/mobility.rb``` like above.

```Marshal.dump(Project.new)``` ok
```Marshal.dump(Project.new(title: "title")``` 
```TypeError (can't dump hash with default proc)```

I think the error is related to ```mobility-1.0.3/lib/mobility/plugins/backend.rb:121``` where you create a  ```::Hash.new do |hash, name|```

**Comment by shioyama** [comment]
Ok, thanks I can indeed reproduce. I don't really know anything about `Marshal.dump` though or what Mobility should do to avoid that problem. Will have to investigate. If you or anyone has info on the correct way to handle this, please let me know (or make a PR to fix! :bow: )

**Comment by shioyama** [comment]
Here's more info on the problem: https://stackoverflow.com/questions/13360541/marshal-cant-dump-hash-with-default-proc-typeerror

**Comment by shioyama** [comment]
Ok I think I understand how to fix this. Basically instead of assigning a `Hash` with a default proc, we'd need to define an actual class which does the same thing but with named method(s), so it can be marshaled.

This shouldn't be hard any anyway maybe makes sense.

**Comment by shioyama** [comment]
@maxime-carbonneau I got a fix that seems to work in #505, have a look.

**Comment by maxime-carbonneau** [comment]
> @maxime-carbonneau I got a fix that seems to work in #505, have a look.

I lightly test it in a real project and it work like a charm.
Thanks for your time

**Comment by shioyama** [comment]
Fixed in #532 and released in [1.2.2](https://github.com/shioyama/mobility/issues/494).
